### ISSUE TITLE
Add dashed lines to the core API

### ISSUE THREAD

**Issue by ianstormtaylor** [processing/p5.js#5670]
### Increasing Access

Unsure. But I believe it would reduce the mental barriers to a very common goal, which feels like increasing access for people with less math-heavy or programming-heavy backgrounds.

### Most appropriate sub-area of p5.js?

- [ ] Accessibility (Web Accessibility)
- [ ] Build tools and processes
- [ ] Color
- [X] Core/Environment/Rendering
- [ ] Data
- [ ] DOM
- [ ] Events
- [ ] Friendly error system
- [ ] Image
- [ ] IO (Input/Output)
- [ ] Localization
- [ ] Math
- [ ] Unit Testing
- [ ] Typography
- [ ] Utilities
- [ ] WebGL
- [ ] Other (specify if possible)

### Feature request details

P5.js should add dashed lines as a core construct. Dashed lines are really common when working in sketches... either for hierarchical reasons (eg. a less important line) or just pure aesthetics reasons (eg. in generative art).

This has been brought up before in https://github.com/processing/p5.js/issues/3016 and https://github.com/processing/p5.js/issues/3336, but it was waived off then.

There is a "workaround" that involves dipping into the Canvas API's own [`setLineDash`](https://developer.mozilla.org/en-US/docs/Web/API/CanvasRenderingContext2D/setLineDash) method, but the problem with that is that you can't rely on dashing across rendering contexts... so if you want to switch to rendering with SVG or WebGL you have to find a new approach. And because of this, downstream plugins and addons can't rely on it.

To have a rendering-agnostic approach right now you'd have to do some very heavy math to be able to convert any potential lines, curves, shapes, etc. to a ton of individual segments, which is not only very complex to write, but likely less performant.

I think this is such common functionality—evidenced by both Canvas and SVG ([`stroke-dasharray`](https://developer.mozilla.org/en-US/docs/Web/SVG/Attribute/stroke-dasharray)) having implemented this natively—that it should be handled in core so that rendering contexts (and external plugins) can build on top of a predictable target.

**Comment by davepagurek** [comment]
> To have a rendering-agnostic approach right now you'd have to do some very heavy math to be able to convert any potential lines, curves, shapes, etc. to a ton of individual segments, which is not only very complex to write, but likely less performant.

I have [an implementation of a fast `getPointAtLength` method for Bezier curves](https://codepen.io/davepvm/pen/mdmPJaa) that I've been [using in p5 sketches](https://openprocessing.org/sketch/1452264) for a little while now. We could potentially use a method like this to do a rendering-agnostic dash. It would just require knowing how to convert primitives to bezier segments. Catmull-Rom curves already get converted to Bezier, straight lines can be converted easily (and maybe can be special cased in `getPointAtLength` for even better performance?) Ellipses will already need to be approximated by polygons in WebGL mode so it's maybe not too bad to convert their outlines into polygons with Bezier tangents smoothing them.

**Comment by golanlevin** [comment]
Respectfully, here's a contrary voice. 

In my book for educators, written with Tega Brain, "Code as Creative Medium", we offer the problem of creating one's own dashed line as a suggested assignment for a beginning programmer. Our feeling was that this is comparatively basic problem that integrates a student's understanding of iteration, some basic mathematics, and a good opportunity to use a function like `lerp()`. 

Personally, I would be a little sad to see this problem "given away" like this. While having a readymade dashed line would "reduce the mental barriers to a very common goal," I think there's no replacement for people _making their own lines_ in a world of readymade, off-the-shelf solutions — most especially for "pure aesthetics" reasons. Providing a readymade dashed line would likely short-circuit a learner from discovering all of the ways in which self-made dashed lines can be *personalized* (for example: by varying the dash duty cycle; producing 1D textures by using dashes of multiple different lengths; randomizing subsegment lengths; futzing with the strokeCap on subsegments, etc.).

There are other ways of achieving lightweight lines (thin strokeWeight; transparency). 

![dashed_line](https://user-images.githubusercontent.com/290053/179134255-3476a94e-85ad-45af-8e4e-a3da427359b8.png)

**Comment by Qianqianye** [comment]
Thank you @ianstormtaylor for sharing the detailed feature request, and @davepagurek and @golanlevin for the comments.

Because we are trying to keep the p5.js API as minimal as possible, the dashed line feature could be a great starting point for an add-on library. Please see the [contributor doc](https://github.com/processing/p5.js/blob/main/contributor_docs/creating_libraries.md) of how to create a library. We'd love to include the library in the [p5js.org/libraries](https://p5js.org/libraries/) page if you decide to make one.

So I will close this issue for now for organizational purpose, and please feel free to re-open it when you'd like to share an update of this issue. Thank you!

**Comment by ianstormtaylor** [comment]
Really disappointing decision IMO—I think P5.js should strive to pave over these rendering layer differences for use cases as common as dashed lines.

@golanlevin I respect your write up, but the reasoning doesn't really make sense to me. It's like arguing that there shouldn't be a `rect` function because you can just use four lines. Or that there shouldn't be `circle` because it would "short-circuit a learner from discovering all of the ways in which [`arc`] can be personalized". 

To me it feels like gatekeeping disguised as "helping people learn". I'd argue it's actually a much more compelling learning experience when you can say "you know this simple built-in method? here's how you can recreate it. and now that you've gone to that trouble, here's what you new power you've unlocked".

So many use cases for dashed lines don't need to dive into the intricacies of how dashes can be rendered (eg. imagine a solar system simulation with solid lines for planetary orbits and dashed lines for lunar orbits). 

The current state of P5.js requires authors to dip into the Canvas API (and then not being able to use that across contexts) or to go down a big rabbit hole of line dashing algorithms just to get an incredibly basic output that most rendering engines and most graphics applications support natively (eg. Photoshop, Illustrator, Figma, Sketch, etc.).

**Comment by golanlevin** [comment]
@ianstormtaylor Thanks for your response. I do **100%** agree with you that dashed lines feel like a "basic" thing.

But I disagree that they're as simple as four-lines-make-a-rect in their actual implementation. Unlike a rect or circle, there's a lot more possible parameters needed to control them (from the front end) and a lot of computational geometry choices to render them (on the back end) — and therefore many more ways to get dashed lines _wrong_. Put another way, I submit that Adobe and Figma can afford to throw a team of developers to make dashed lines feel "basic".

In my opinion, I think there's enough subtlety (and enough different points of view) regarding how dashed lines should be "properly" implemented that it would be best to _begin with_ an external library that offers up a possible implementation. Perhaps something similar in operation to [p5.scribble.js](https://github.com/generative-light/p5.scribble.js). In your response, you didn't explicitly address this library suggestion, and I'd be interested to hear your thoughts about this. There's plenty of case history for add-on libraries becoming absorbed into the main body of toolkits, as you probably know.

I don't know enough about how lines/curves are implemented throughout p5. But unless it's trivial to swap in lines drawn with the Canvas-API `setLineDash` method, I'm concerned that a official implementation made from scratch will open a Pandora's box of complaints—because (for example) it is inefficient, or because it doesn't provide fine-grained control of dash-versus-gap lengths, or because it handles curves and polylines poorly, or because it allows for duty-cycles that don't force a dash on the final point, or because it produces ugly results for ellipses (whose perimeters are [non-trivial to calculate](https://www.mathsisfun.com/geometry/ellipse-perimeter.html)), or because the resampled dashes flicker distractingly in response to small coordinate changes, or because the dashes are locked to an invisible lattice (as they were in OpenGLv1).

I regret the way I phrased my reply earlier. I think that making dashed lines is something that _feels_ comparatively basic (and can be learnfully implemented by a bright student to meet their own needs). But this elides a lot of complexity about their actual implementation, and making a solution to meet _everyone's_ needs (including graphics professionals) is another beast entirely.

**Comment by ianstormtaylor** [comment]
@golanlevin I understand your response, and I agree with you that there are many different ways that dashed lines can be complex-ly defined… People could want every other dash to be rounded, or every fifth dash to be slightly translucent, or dashes to maintain a consistent gap while rounding a curve, etc., etc., etc.

That doesn't mean P5.js needs to throw out the simple case, which solves 90% (or more) of applications, such that beginners are effectively blocked from quickly adding dashed lines to a sketch.

> @golanlevin: I don't know enough about how lines/curves are implemented throughout p5. But unless it's trivial to swap in lines drawn with the Canvas-API setLineDash method, I'm concerned that a official implementation made from scratch will open a Pandora's box of complaints…

It should be trivial I believe, yes. It would delegate to Canvas's underlying APIs just like for regular lines and other stroke-related methods.

Other examples of handling these simpler cases... P5 doesn't tell you that to achieve rounded line caps you have to use your own circles. Or to achieve beveled joins you have to mask them yourself. These are all basic stroke primitives.

---

### Examples
This problem is extremely table stakes, evidenced by how often it's implemented in drawing contexts.

#### Canvas
Here's how Canvas solves it…

```js
ctx.beginPath();
ctx.setLineDash([5, 15]);
ctx.moveTo(0, 50);
ctx.lineTo(300, 50);
ctx.stroke();
```

#### SVG
And here's how SVG does it (using the same `dash + gap` API design):

```js
<line 
  x1="0" y1="5" 
  x2="30" y2="5" 
  stroke="black"
  stroke-dasharray="4 1" 
/>
```

#### Figma
Here are the dashed line controls in Figma (same `dash + gap` API again):

<img width="269" alt="image" src="https://user-images.githubusercontent.com/311752/189550262-42dffe87-26c9-4ac8-b440-70981d88aab3.png">

#### Sketch
Here they are in Sketch (same thing):

<img width="188" alt="image" src="https://user-images.githubusercontent.com/311752/189550278-cc67b521-f356-4169-aec5-5e62d0d68447.png">

#### Illustrator
Here is Illustrator (same again):

<img width="240" alt="image" src="https://user-images.githubusercontent.com/311752/189550315-86664e82-bc07-4084-a760-a0363c206ae3.png">

#### Inkscape
Here it is in Inkscape (same again, and recently added):

<img width="280" alt="image" src="https://user-images.githubusercontent.com/311752/189550685-3193ef24-3288-4956-8687-3834290c5bf9.png">

This is a problem that every vector editing program deals with, and both Canvas and SVG deal with. And they all use a simple but flexible API of `dash + gap` values with an optional dash `offset`.

---

### Implementation
I think it should be implemented in P5.js as:

```js
strokeDash(10, 5)
strokeDashOffset(-5) // optional, sets the point at which to start dashes
line(0, 0, 100, 100)
```

This mimics its other stroke-related APIs which often just call directly into Canvas APIs like `strokeCap -> lineCap` or `strokeJoin -> lineJoin`

Which would be equivalent to Canvas's:

```js
ctx.beginPath()
ctx.setLineDash([10, 5])
ctx.lineDashOffset = -5
ctx.moveTo(0, 0)
ctx.lineTo(100, 100)
ctx.stroke()
```

And SVG's:

```js
<line 
  x1="0" y1="0"
  x2="100" y2="100"
  stroke="black"
  stroke-dasharray="10 5"
  stroke-dashoffset="-5"
/>
```

**Comment by golanlevin** [comment]
>  It would delegate to Canvas's underlying APIs just like for regular lines and other stroke-related methods.

If p5 can indeed delegate dashed lines to the Canvas API, then (FWIW) I've come around and agree that you're right that dashed lines should eventually be implemented into the core. Your proposed API makes sense, too. 

Questions remain about the extent to which the p5 WebGL needs to mirror the Canvas implementation. Would its dashes need to resemble the Canvas dashes _exactly_? I'm also having a hard time even finding an official WebGL solution for dashed lines (though three.js has an [attractive demo](https://threejs.org/examples/webgl_lines_dashed.html)).

**Comment by davepagurek** [comment]
Just throwing in my two cents about the technical side of things: I think delegating to the Canvas API in 2D mode should be feasible as I believe we are already drawing everything in one `beginPath()`/`stroke()`.

WebGL lines already don't mirror 2D, so I think matching exactly isn't doesn't need to prevent adding this if it's plausible we can get something similar working eventually. It definitely isn't *trivial* to add to WebGL, but I've been looking into getting stroke caps/joins to work more like in 2D mode, and the potential of having dashes can fit into the design. Everything gets converted to line segments right now anyway so it won't take much math to pass the distance along the line in as a vertex attribute, and we can maybe use a fragment shader to hide parts of the dashes. tl;dr it's a lot easier to do this in 2D mode but it's still plausible in WebGL longer term.

**Comment by ianstormtaylor** [comment]
@golanlevin thanks for keeping an open mind! I really appreciate it.

**Comment by golanlevin** [comment]
@ianstormtaylor you bet. Though I'm still not sure what happens with more complex types of graphics, such as polylines, when dashing is delegated to Canvas....

**Comment by limzykenneth** [comment]
I'm reopening this for further discussion from other members of the community and the relevant area stewards. Feel free to hash out some implementation details such as defaults and customisability. 

One note from me is that it is good to keep in mind that p5.js don't have to be the first and last thing a beginner coder learn and use, I'd personally much prefer that p5.js can gently lead learners to further explore the rest of the ecosystem and programming in general, even if that means they use p5.js less in their future as a result.

**Comment by alvinometric** [comment]
A year later, I still think this is a great idea fwiw

**Comment by sudhanshuv1** [comment]
If we still prefer to add this feature, I would like to work on this issue. But I couldn't find any easy to understand implementation of dashed lines in GLSL.

**Comment by davepagurek** [comment]
Here's a quick sketch with a drawing app that does dashed lines: https://editor.p5js.org/davepagurek/sketches/r6QxdNb5t This GLSL approach is a bit basic and would take extra effort to support caps on the dashes, but could be a starting point.

**Comment by limzykenneth** [comment]
I'll try to explore this in an addon library for 2D mode and possibly use it to figure out the library interface for 2.0.

**Comment by GregStanton** [comment]
Just chiming in here with a quick note: I think we may be able to do a renderer-agnostic implementation now, following @davepagurek's [comment](https://github.com/processing/p5.js/issues/5670#issuecomment-1175016804). Now that we have the infrastructure in place, the idea would be to create a `PointAtLengthGetter` visitor class, which would complement the existing `Shape` visitor classes. 

If we go that route, it may be nice to finish cleaning up the Custom Shapes submodule and to add some internal documentation. Then if another contributor wants to do the implementation, it'd be a good test of how accessible this part of the codebase is; improving accessibility for conributors was one of our design goals.

**Comment by NalinDalal** [comment]
@ksen0 @davepagurek is this open, can i try this, will update in few days??
not going hardly for it, just trying out new issue

**Comment by ksen0** [comment]
Hi @NalinDalal , thanks for the interest, as written in the last comment - please first propose a specific API here as a comment, rather than working directly. Some API ideas and important comments have already been proposed, so please consider these if you're interested.

**Comment by NalinDalal** [comment]
Hi, since the other issue I have asked for grabs is blocked by some other pull request, and is being currently worked on, I would love to have a look at this

**Comment by NalinDalal** [comment]
## Proposed API

### Functions

```js
strokeDash(pattern)
strokeDashOffset(offset)
```

### Parameters

**pattern**: `number[]`  
Defines alternating dash and gap lengths in pixels.

**offset**: `number`  
Defines the starting offset along the stroke.

---

## Examples

### Basic dashed line
```js
strokeDash([8, 4])
line(0, 0, 100, 0)
```

### Complex dash pattern
```js
strokeDash([10, 4, 2, 4])
ellipse(50, 50, 80, 80)
```

### Offset dash phase
```js
strokeDash([6, 3])
strokeDashOffset(5)
line(0, 0, 200, 0)
```

### Reset to solid stroke
```js
strokeDash([])
line(0, 0, 100, 0)
```

---

## Design Goals

### Consistency
Align with existing stroke styling APIs:

- `strokeWeight()`
- `strokeCap()`
- `strokeJoin()`

Maintain familiar naming patterns and behavior.

---

### Predictability
Behavior mirrors established graphics standards:

- Canvas: `CanvasRenderingContext2D.setLineDash`
- SVG: `stroke-dasharray`

Users coming from other creative coding or design tools should find the behavior intuitive.

---

### Extensibility
The array-based pattern enables:

- dotted lines → `[1, 4]`
- dash-dot patterns → `[10, 4, 2, 4]`
- custom rhythmic strokes
- procedural stroke styling in future addons

The API avoids constraining future stroke styling features.

---

## Renderer Behavior

### 2D Renderer
Delegates to Canvas API when available:

```js
drawingContext.setLineDash(pattern)
drawingContext.lineDashOffset = offset
```

Provides performance and visual parity with native browser rendering.

---

### WebGL Renderer

Initial implementation may approximate dash patterns via polyline segmentation.

Future implementations may use shader-based approaches:

- distance-along-path attribute
- fragment shader masking
- improved cap handling

Exact visual parity with 2D mode is desirable but not required for initial release.

---

## Fallback Behavior

```js
strokeDash([])
```

Resets stroke style to solid.

Default state:

```js
strokeDash([])
strokeDashOffset(0)
```

---

## Open Questions

### Stroke scaling
Should dash lengths scale with `strokeWeight()`?

Example:
```js
strokeWeight(4)
strokeDash([4, 2])
```

Should this visually scale proportionally?

---

### Continuity across shapes
Should dash phase persist across multiple drawing calls?

Example:
```js
strokeDash([8, 4])
line(0, 0, 100, 0)
line(100, 0, 200, 0)
```

Should the second line continue the dash rhythm?

---

### beginShape behavior
Should dash continuity persist across vertices?

Example:
```js
beginShape()
vertex(0, 0)
vertex(100, 0)
vertex(100, 100)
endShape()
```

Expected behavior needs clarification.

---

### Offset semantics
Should offset be:

- global state
- reset per shape
- reset per draw loop

---

## Prior Art

Canvas:
```js
ctx.setLineDash([5, 15])
ctx.lineDashOffset = -5
```

SVG:
```xml
<line
  x1="0"
  y1="0"
  x2="100"
  y2="0"
  stroke="black"
  stroke-dasharray="5 15"
  stroke-dashoffset="-5"
/>
```

Design tools supporting similar models:

- Figma
- Illustrator
- Sketch
- Inkscape

---

## Summary

This proposal introduces dashed stroke styling as a first-class primitive in p5.js, aligning with existing stroke APIs while remaining renderer-agnostic and extensible for future improvements.

**Comment by NalinDalal** [comment]
## Proposed API

### Functions

```js
strokeDash(pattern)
strokeDashOffset(offset)
```

### Parameters

**pattern**: `number[]`  
Defines alternating dash and gap lengths in pixels.

**offset**: `number`  
Defines the starting offset along the stroke.

---

## Examples

### Basic dashed line
```js
strokeDash([8, 4])
line(0, 0, 100, 0)
```

### Complex dash pattern
```js
strokeDash([10, 4, 2, 4])
ellipse(50, 50, 80, 80)
```

### Offset dash phase
```js
strokeDash([6, 3])
strokeDashOffset(5)
line(0, 0, 200, 0)
```

### Reset to solid stroke
```js
strokeDash([])
line(0, 0, 100, 0)
```

---

## Design Goals

### Consistency
Align with existing stroke styling APIs:

- `strokeWeight()`
- `strokeCap()`
- `strokeJoin()`

Maintain familiar naming patterns and behavior.

---

### Predictability
Behavior mirrors established graphics standards:

- Canvas: `CanvasRenderingContext2D.setLineDash`
- SVG: `stroke-dasharray`

Users coming from other creative coding or design tools should find the behavior intuitive.

---

### Extensibility
The array-based pattern enables:

- dotted lines → `[1, 4]`
- dash-dot patterns → `[10, 4, 2, 4]`
- custom rhythmic strokes
- procedural stroke styling in future addons

The API avoids constraining future stroke styling features.

---

## Renderer Behavior

### 2D Renderer
Delegates to Canvas API when available:

```js
drawingContext.setLineDash(pattern)
drawingContext.lineDashOffset = offset
```

Provides performance and visual parity with native browser rendering.

---

### WebGL Renderer

Initial implementation may approximate dash patterns via polyline segmentation.

Future implementations may use shader-based approaches:

- distance-along-path attribute
- fragment shader masking
- improved cap handling

Exact visual parity with 2D mode is desirable but not required for initial release.

---

## Fallback Behavior

```js
strokeDash([])
```

Resets stroke style to solid.

Default state:

```js
strokeDash([])
strokeDashOffset(0)
```

---

## Open Questions

### Stroke scaling
Should dash lengths scale with `strokeWeight()`?

Example:
```js
strokeWeight(4)
strokeDash([4, 2])
```

Should this visually scale proportionally?

---

### Continuity across shapes
Should dash phase persist across multiple drawing calls?

Example:
```js
strokeDash([8, 4])
line(0, 0, 100, 0)
line(100, 0, 200, 0)
```

Should the second line continue the dash rhythm?

---

### beginShape behavior
Should dash continuity persist across vertices?

Example:
```js
beginShape()
vertex(0, 0)
vertex(100, 0)
vertex(100, 100)
endShape()
```

Expected behavior needs clarification.

---

### Offset semantics
Should offset be:

- global state
- reset per shape
- reset per draw loop

---

## Prior Art

Canvas:
```js
ctx.setLineDash([5, 15])
ctx.lineDashOffset = -5
```

SVG:
```xml
<line
  x1="0"
  y1="0"
  x2="100"
  y2="0"
  stroke="black"
  stroke-dasharray="5 15"
  stroke-dashoffset="-5"
/>
```

Design tools supporting similar models:

- Figma
- Illustrator
- Sketch
- Inkscape

---

## Summary

This proposal introduces dashed stroke styling as a first-class primitive in p5.js, aligning with existing stroke APIs while remaining renderer-agnostic and extensible for future improvements.

**Comment by ksen0** [comment]
Hi @NalinDalal , thanks for looking into this. (I hid one of the comments as they appeared to be duplicates? Pelase let me kow otherwise.)

A couple of thoughts:

> Should dash phase persist across multiple drawing calls?

Other similar functions (strokeWight, fill, stroke) would. It's possible to have no stroke or no fill.  Maybe something like `strokeDash(NONE)` would make sense?

I am a bit conflicted about a separate strokeDashOffset function, rather than an optional parameter (like `strokeDash([3,4])` having 0 offset, and `strokeDash([3,4],2)` having offset of 2)`.

> Should dash lengths scale with strokeWeight()?

In my opinion, it would be good to match the behavior of how dashed lines behave in 2D with the drawing context method. One way to do this:

```
function setup() {
  createCanvas(300, 300);
  noLoop();
}

function draw() {
  background(220);
  
  strokeWeight(1);
  drawingContext.setLineDash([10]);
  rect(50, 50, 200, 200);
  
  strokeWeight(6);
  drawingContext.setLineDash([10]);
  rect(100, 100, 100, 100);
}
```

So the lengths there do not change based on weight.

Although I agree total visual parity with 2D is not required initially, I think it's very important that any proposed API expansion does work _somehow_ in WEBGL: as in, creates some kind of dashed line, and uses the same function signatures to do so.

I think the best way forward, like is some of the other issues that propose API additions, is to implement the proof of concept as a p5.js 2.x add-on (`npm create p5.js-addon`) for community testing of the API. (You're welcome to release it separately as an add-on and then propose it into core, but the main point at the moment is to make it somehow available for testing and discussion before considering merging into the core library.)

**Comment by NalinDalal** [comment]
## Revised Proposed API

### Function

```js
strokeDash(pattern, offset?)
```

### Parameters

**pattern**: `number[] | NONE`  
Defines alternating dash and gap lengths in pixels.

Examples:

```js
[8, 4]        // dash-gap
[10, 4, 2, 4] // dash-gap-dot-gap
[1, 3]        // dotted
```

Passing `NONE` or an empty array resets to a solid stroke:

```js
strokeDash(NONE)
strokeDash([])
```

---

**offset (optional)**: `number`  
Defines the starting offset along the stroke.

Default:

```js
offset = 0
```

Example:

```js
strokeDash([6, 3], 2)
```

---

## Examples

### Basic dashed line

```js
strokeDash([8, 4])
line(0, 0, 100, 0)
```

---

### Offset dash phase

```js
strokeDash([6, 3], 5)
line(0, 0, 200, 0)
```

---

### Complex pattern

```js
strokeDash([10, 4, 2, 4])
ellipse(50, 50, 80, 80)
```

---

### Reset to solid stroke

```js
strokeDash(NONE)
line(0, 0, 100, 0)
```

---

## Behavior

### Persistent state

Like other stroke styling functions:

```js
stroke()
strokeWeight()
strokeCap()
strokeJoin()
```

dash settings persist across draw calls until changed.

Example:

```js
strokeDash([8, 4])

line(0, 0, 100, 0)
rect(20, 20, 60, 60)
ellipse(50, 50, 40, 40)
```

All shapes use the same dash pattern.

---

### Interaction with strokeWeight

Dash/gap lengths are **not scaled** by stroke weight.

This matches Canvas behavior:

```js
strokeWeight(1)
strokeDash([10])
rect(50, 50, 200, 200)

strokeWeight(6)
strokeDash([10])
rect(100, 100, 100, 100)
```

Both rectangles use the same dash length.

---

## Renderer behavior

### 2D mode

Delegates to Canvas API:

```js
drawingContext.setLineDash(pattern)
drawingContext.lineDashOffset = offset
```

Provides predictable and performant behavior consistent with existing web standards.

---

### WEBGL mode

Initial implementation goal:

Produce visually dashed strokes using the same function signature.

Possible approaches:

• polyline segmentation based on arc length  
• distance-along-path vertex attribute  
• fragment shader masking  

Exact visual parity with 2D mode is not required initially, but the API must function consistently across renderers.

---

## Continuity rules

Dash phase continues across drawing calls within the same shape.

Example:

```js
strokeDash([8, 4])

line(0, 0, 100, 0)
line(100, 0, 200, 0)
```

Second line continues the dash rhythm.

---

For shape primitives:

```js
beginShape()
vertex(0, 0)
vertex(100, 0)
vertex(100, 100)
endShape()
```

Dash spacing should remain continuous across vertices.

---

## Design goals

### Consistency

Follows established stroke styling patterns:

```js
strokeWeight()
strokeCap()
strokeJoin()
```

Maintains familiar state-based styling model.

---

### Predictability

Aligns with:

Canvas → `setLineDash`
SVG → `stroke-dasharray`

Users transitioning from design tools or browser APIs encounter minimal surprise.

---

### Extensibility

Pattern array allows future support for:

• procedural strokes  
• textured strokes  
• animated dash offsets  
• brush-style rendering  
• addon experimentation  

---

## Addon-first approach

As suggested, the initial implementation can be developed as a **p5.js 2.x addon**:

```bash
npm create p5.js-addon
```

Benefits:

• enables API testing without core constraints  
• allows iteration on WEBGL behavior  
• gathers community feedback  
• validates ergonomics  
• reduces risk of premature core expansion  

If the API proves stable, the implementation can later be proposed for inclusion in core.

---

## Summary

This proposal introduces dashed stroke styling as a persistent stroke state, consistent with existing p5 APIs and compatible with both 2D and WEBGL renderers.

The API prioritizes:

• familiarity  
• minimal surface area  
• renderer-agnostic behavior  
• future extensibility

**Comment by ksen0** [comment]
Hi @NalinDalal , the content of this approach makes sense. Dashed settings probably should also be included in push/pop state.

Minor note: To me the style and formatting of your comments appears to be AI-generated. Please note that AI PRs are not allowed in the [p5.js AI usage policy](https://github.com/processing/p5.js/blob/main/AI_USAGE_POLICY.md) ("This project does not accept fully AI-generated pull requests. AI tools may be used assistively only. You must understand and take responsibility for every change you submit."). You're welcome to work on the above proposal, and I'm glad you've brought this topic back up! I just wanted to flag that it's important to be mindful of not overusing LLMs. Thanks you for understanding.

**Comment by NalinDalal** [comment]
Thanks for your insight, I totally understand your concerns over the use of LLM, it's just that i wanted to format the comment, hence I had used LLM, rest assured the approach and engineering will always be mine

**Comment by NalinDalal** [comment]
@ksen0 can I proceed to work on this then?

**Comment by NalinDalal** [comment]
Update: Since the other issue I had asked originally is now ready to be worked on and isn't blocked by ongoing pr, I will hold this one, If anyone else wants to work on this, they are free to work on it and i would be happy to help them around.
